token-metadata/program/src/processor.rs

```diff
@@ -1840,6 +1846,232 @@ pub fn process_burn_nft(program_id: &Pubkey, accounts: &[AccountInfo]) -> Progra
+
+    let edition_lamports = print_edition_info.lamports();
+    **print_edition_info.try_borrow_mut_lamports()? = 0;
+    **owner_info.try_borrow_mut_lamports()? = owner_info
+        .lamports()
+        .checked_add(edition_lamports)
+        .ok_or(MetadataError::NumericalOverflowError)?;
+
+    let metadata_data = &mut metadata_info.try_borrow_mut_data()?;
+    let edition_data = &mut print_edition_info.try_borrow_mut_data()?;
+    let edition_data_len = edition_data.len();
+
+    // Use MAX_METADATA_LEN because it has unused padding, making it longer than current metadata len.
+    sol_memset(metadata_data, 0, MAX_METADATA_LEN);
+    sol_memset(edition_data, 0, edition_data_len);
+
+    //       **EDITION HOUSEKEEPING**
+    // Set the particular bit for this edition to 0 to allow reprinting,
+    // IF the print edition owner is also the master edition owner.
+    // Otherwise leave the bit set to 1 to disallow reprinting.
+    let mut edition_marker: EditionMarker = EditionMarker::from_account_info(edition_marker_info)?;
+    let owner_is_the_same = *owner_info.key == master_edition_token_account.owner;
+
+    if owner_is_the_same {
+        let (index, mask) = EditionMarker::get_index_and_mask(print_edition.edition)?;
+        edition_marker.ledger[index] ^= mask;
+    }
+
+    // If the entire edition marker is empty, then we can close the account.
+    // Otherwise, serialize the new edition marker and update the account data.
+    if edition_marker.ledger.iter().all(|i| *i == 0) {
+        let edition_marker_lamports = edition_marker_info.lamports();
+        **edition_marker_info.try_borrow_mut_lamports()? = 0;
+        **owner_info.try_borrow_mut_lamports()? = owner_info
+            .lamports()
+            .checked_add(edition_marker_lamports)
+            .ok_or(MetadataError::NumericalOverflowError)?;
+
+        let edition_marker_data = &mut edition_marker_info.try_borrow_mut_data()?;
+        let edition_marker_data_len = edition_marker_data.len();
+
+        sol_memset(edition_marker_data, 0, edition_marker_data_len);
+    } else {
+        let mut edition_marker_info_data = edition_marker_info.try_borrow_mut_data()?;
+        edition_marker_info_data[0..].fill(0);
+        edition_marker.serialize(&mut *edition_marker_info_data)?;
+    }
+
+    // Decrement the suppply on the master edition now that we've successfully burned a print.
+    // Decrement max_supply if Master Edition owner is not the same as Print Edition owner.
+    let mut master_edition: MasterEditionV2 =
+        MasterEditionV2::from_account_info(master_edition_info)?;
+    master_edition.supply = master_edition
+        .supply
+        .checked_sub(1)
+        .ok_or(MetadataError::NumericalOverflowError)?;
+
+    if let Some(max_supply) = master_edition.max_supply {
+        if !owner_is_the_same {
+            master_edition.max_supply = Some(
+                max_supply
+                    .checked_sub(1)
```

Comment:
Would there be any benefit to trying to add a test that tries to decrement these supply/max supply values below 0?  I suspect it might not be a possible scenario, but maybe future proofing for catching some bug like if someone removes the `checked_sub` or makes some other code changes.

Reply:
They are both `u64`'s so we'd be testing Rust's primitive integer types. The only reason to have a test would be if we both changed the types to unsigned integers (a breaking change) *and* removed the checked_sub. I think that's probably not worth worrying about at this point?

Reply:
yeah it is probably fine as is, if someone was changing the types and making the subtraction unsafe both we should be able to detect it in PR.